proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,417 @@
+
+   To see why the above algorithm is safe, assuming one block X is
+   optimistically confirmed before the restart, we prove that it will always be
+   either the block picked or the ancestor of the block picked.
+
+   Assume X is not picked, then:
+
+   1. If its parent Y is on the Heaviest fork, then either a sibling of X is
+   chosen or no child of Y is chosen. In either case
+   vote_on_X + stake_on_validators_not_in_restart < 62%, otherwise X would be
+   picked. This contradicts with the fact X got 67% votes before the restart,
+   because vote_on_X should have been greater than
+   67% - 5% (non-conforming) - stake_on_validators_not_in_restart.
+
+   2. If its parent Y is also not on the Heaviest fork, all ancestors of X
+   should have > 67% of the votes before restart. We should be able to find
+   the last ancestor of X on the Heaviest fork, then contradiction in previous
+   paragraph applies.
+
+   In some cases, we might pick a child with less than 67% votes before the
+   restart. Say a block X has child A with 43% votes and child B with 42%
+   votes, child A will be picked as the restart slot. Note that block X which
+   has more than 43% + 42% = 85% votes is the ancestor of the picked restart
+   slot, so the constraint that optimistically confirmed block never gets
+   rolled back is satisfied.
+
+   After deciding heaviest block, gossip
+   `RestartHeaviestFork(X.slot, X.hash, committed_stake_percent)` out, where X
+   is the latest picked block. We also gossip stake of received
+   `RestartHeaviestFork` messages so that we can proceed to next step when
+   enough validators are ready.
+
+### Exit `wen restart phase`
+
+**Restart if everything okay, halt otherwise**
+
+The main purpose in this step is to decide the `cluster restart slot` and the
+actual block to restart from.
+
+All validators in restart keep counting the number of `RestartHeaviestFork`
+where `received_heaviest_stake` is higher than `RESTART_STAKE_THRESHOLD`. Once
+a validator counts that `RESTART_STAKE_THRESHOLD` of the validators send out
+`RestartHeaviestFork` where `received_heaviest_stake` is higher than
+`RESTART_STAKE_THRESHOLD`, it starts the following checks:
+
+* Whether all `RestartHeaviestFork` have the same slot and same bank Hash.
+Because validators are only sending slots instead of bank hashes in 
+`RestartLastVotedForkSlots`, it's possible that a duplicate block can make the
+cluster unable to reach consensus. So bank hash needs to be checked as well.
+
+* The voted slot is equal or a child of local optimistically confirmed slot.
+
+If all checks pass, the validator immediately starts setting root and
+generating an incremental snapshot at the agreed upon `cluster restart slot`.
+
+After the snapshot generation is complete, and above checks pass, it then
+proceeds to issue a hard fork at the designated slot and change shred version
+in gossip. After that it restarts into the normal (non-restart) state.
```

Comment:
We should issue the hard fork upon setting the root, so the hard fork is included in the incremental snapshot. This will ensure that if the validator dies after this step and is restarted normally, it will follow the hard fork

Reply:
Changed, does this look better now?